Description: Carry Ubuntu specific machine types

Since Ubuntu is a downstream of qemu carrying patches it needs custom machine
types to be able to identify and manage the delta that might affect machine
types.

This is an important piece to keep cross release migration supported for any
downstream.

Since the p->t transition these types are mostly stable copies of the upstream
type (in the past this was more unstable upstream, so there was more delta),
but they need to stay specific to reflect the delta we have. And even more so
to have something to base off for affecting SRU changes.

Package maintainers please see https://wiki.ubuntu.com/QemuKVMMigration when
maintaining this patch on SRU, merge or other packaging activity.

Original-Author: Serge Hallyn <serge.hallyn@ubuntu.com>
Original-Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1304107
Author: Christian Ehrhardt <christian.ehrhardt@canonical.com>
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1621042
Forwarded: no

--- a/hw/i386/pc_piix.c
+++ b/hw/i386/pc_piix.c
@@ -473,7 +473,7 @@
 {
     pc_i440fx_machine_options(m);
     m->alias = "pc";
-    m->is_default = 1;
+    m->is_default = 0;
 }
 
 DEFINE_I440FX_MACHINE(v2_5, "pc-i440fx-2.5", NULL,
@@ -1057,3 +1057,54 @@
 DEFINE_PC_MACHINE(xenfv, "xenfv", pc_xen_hvm_init,
                   xenfv_machine_options);
 #endif
+
+/* Ubuntu machine types */
+static void pc_trusty_machine_options(MachineClass *m)
+{
+    pc_i440fx_machine_options(m);
+    m->desc = "Ubuntu 14.04 PC (i440FX + PIIX, 1996)";
+}
+DEFINE_I440FX_MACHINE(trusty, "pc-i440fx-trusty", pc_compat_2_0,
+                      pc_trusty_machine_options)
+
+static void pc_utopic_machine_options(MachineClass *m)
+{
+    pc_i440fx_machine_options(m);
+    m->desc = "Ubuntu 14.10 PC (i440FX + PIIX, 1996)";
+    m->default_display = "std";
+}
+DEFINE_I440FX_MACHINE(utopic, "pc-i440fx-utopic", pc_compat_2_0,
+                      pc_utopic_machine_options)
+
+static void pc_vivid_machine_options(MachineClass *m)
+{
+    pc_i440fx_2_3_machine_options(m);
+    pc_i440fx_machine_options(m);
+    m->desc = "Ubuntu 15.04 PC (i440FX + PIIX, 1996)",
+    m->default_display = "std";
+}
+
+DEFINE_I440FX_MACHINE(vivid, "pc-i440fx-vivid", pc_compat_2_3,
+                      pc_vivid_machine_options);
+
+static void pc_wily_machine_options(MachineClass *m)
+{
+    pc_i440fx_2_4_machine_options(m);
+    pc_i440fx_machine_options(m);
+    m->desc = "Ubuntu 15.04 PC (i440FX + PIIX, 1996)",
+    m->default_display = "std";
+}
+
+DEFINE_I440FX_MACHINE(wily, "pc-i440fx-wily", pc_compat_2_3,
+                      pc_wily_machine_options);
+
+static void pc_xenial_machine_options(MachineClass *m)
+{
+    pc_i440fx_2_5_machine_options(m);
+    m->desc = "Ubuntu 16.04 PC (i440FX + PIIX, 1996)",
+    m->alias = "ubuntu";
+    m->is_default = 1;
+}
+
+DEFINE_I440FX_MACHINE(xenial, "pc-i440fx-xenial", NULL,
+                      pc_xenial_machine_options);
--- a/hw/i386/pc_q35.c
+++ b/hw/i386/pc_q35.c
@@ -485,3 +485,14 @@
 
 DEFINE_Q35_MACHINE(v1_4, "pc-q35-1.4", pc_compat_1_4,
                    pc_q35_1_4_machine_options);
+
+/* Ubuntu machine types */
+static void pc_q35_xenial_machine_options(MachineClass *m)
+{
+    pc_q35_2_5_machine_options(m);
+    m->desc = "Ubuntu 16.04 PC (Q35 + ICH9, 2009)";
+    /* The ubuntu alias and default is on a i440fx type */
+    m->alias = NULL;
+}
+DEFINE_Q35_MACHINE(xenial, "pc-q35-xenial", NULL,
+                   pc_q35_xenial_machine_options);
--- a/hw/ppc/spapr.c
+++ b/hw/ppc/spapr.c
@@ -2439,8 +2439,6 @@
 
     mc->name = "pseries-2.5";
     mc->desc = "pSeries Logical Partition (PAPR compliant) v2.5";
-    mc->alias = "pseries";
-    mc->is_default = 1;
     smc->dr_lmb_enabled = true;
 }
 
@@ -2450,6 +2448,25 @@
     .class_init    = spapr_machine_2_5_class_init,
 };
 
+/* Ubuntu machine types */
+static void spapr_machine_xenial_class_init(ObjectClass *oc, void *data)
+{
+    MachineClass *mc = MACHINE_CLASS(oc);
+    sPAPRMachineClass *smc = SPAPR_MACHINE_CLASS(oc);
+
+    mc->name = "pseries-xenial";
+    mc->desc = "Ubuntu 16.04 pSeries Logical Partition (PAPR compliant)";
+    mc->alias = "pseries";
+    mc->is_default = 1;
+    smc->dr_lmb_enabled = true;
+}
+
+static const TypeInfo spapr_machine_xenial_info = {
+    .name          = MACHINE_TYPE_NAME("pseries-xenial"),
+    .parent        = TYPE_SPAPR_MACHINE,
+    .class_init    = spapr_machine_xenial_class_init,
+};
+
 static void spapr_machine_register_types(void)
 {
     type_register_static(&spapr_machine_info);
@@ -2458,6 +2475,7 @@
     type_register_static(&spapr_machine_2_3_info);
     type_register_static(&spapr_machine_2_4_info);
     type_register_static(&spapr_machine_2_5_info);
+    type_register_static(&spapr_machine_xenial_info);
 }
 
 type_init(spapr_machine_register_types)
--- a/hw/s390x/s390-virtio-ccw.c
+++ b/hw/s390x/s390-virtio-ccw.c
@@ -297,9 +297,7 @@
 {
     MachineClass *mc = MACHINE_CLASS(oc);
 
-    mc->alias = "s390-ccw-virtio";
     mc->desc = "VirtIO-ccw based S390 machine v2.5";
-    mc->is_default = 1;
 }
 
 static const TypeInfo ccw_machine_2_5_info = {
@@ -308,11 +306,28 @@
     .class_init    = ccw_machine_2_5_class_init,
 };
 
+/* Ubuntu machine types */
+static void ccw_machine_xenial_class_init(ObjectClass *oc, void *data)
+{
+    MachineClass *mc = MACHINE_CLASS(oc);
+
+    mc->alias = "s390-ccw-virtio";
+    mc->desc = "Ubuntu 16.04 VirtIO-ccw based S390 machine";
+    mc->is_default = 1;
+}
+
+static const TypeInfo ccw_machine_xenial_info = {
+    .name          = MACHINE_TYPE_NAME("s390-ccw-virtio-xenial"),
+    .parent        = TYPE_S390_CCW_MACHINE,
+    .class_init    = ccw_machine_xenial_class_init,
+};
+
 static void ccw_machine_register_types(void)
 {
     type_register_static(&ccw_machine_info);
     type_register_static(&ccw_machine_2_4_info);
     type_register_static(&ccw_machine_2_5_info);
+    type_register_static(&ccw_machine_xenial_info);
 }
 
 type_init(ccw_machine_register_types)
